server2/Arcadia.Assistant/Arcadia.Assistant.Sharepoint/SharepointSynchronizator.cs

```diff
@@ -0,0 +1,170 @@
+    {
+        #region ctor
+
+        public SharepointSynchronizator(ISickLeaves sickLeaves, IVacations vacations, IWorkHoursCredit workouts, ISharepointDepartmentsCalendarsSettings departmentsCalendarsSettings, ILogger? logger)
+        {
+            this.sickLeaves = sickLeaves;
+            this.vacations = vacations;
+            this.workouts = workouts;
+            this.departmentsCalendarsSettings = departmentsCalendarsSettings;
+            this.logger = logger;
+        }
+
+        #endregion
+
+        #region public interface
+
+        public async Task Synchronize(IEmployees employees, IEnumerable<string> departments, IExternalStorage storage, CancellationToken cancellationToken)
+        {
+            var vacationsSync = new EmployeeVacationsSynchronization(storage, this.logger);
+            var workoutsSync = new EmployeeWorkoutsSynchronization(storage, this.logger);
+            var sickLeavesSync = new EmployeeSickLeavesSynchronization(storage, this.logger);
+
+            var sharepointItemsByCalendars = departments.Distinct().ToDictionary(x => x,
+                dId => this.GetSharepointCalendarsByDepartment(dId).ToDictionary(x => x,
+                    async cal => await this.GetAllSharepointItemsForCalendar(storage, cal)));
+
+            foreach (var departmentId in departments)
+            {
+                var departmentEmployees = await employees.FindEmployeesAsync(EmployeesQuery.Create().ForDepartment(departmentId), cancellationToken);
+                var employeeIds = departmentEmployees.Select(x => x.EmployeeId).ToArray();
+
+                var employeeVacationsTask = this.vacations.GetCalendarEventsByEmployeeAsync(employeeIds, cancellationToken);
+                var employeeWorkoutsTask = this.workouts.GetCalendarEventsByEmployeeMapAsync(employeeIds, cancellationToken);
+                var employeeSickLeavesTask = this.sickLeaves.GetCalendarEventsByEmployeeMapAsync(employeeIds, cancellationToken);
+                await Task.WhenAll(employeeVacationsTask, employeeWorkoutsTask, employeeSickLeavesTask);
+
+                var employeeVacationValues = employeeVacationsTask.Result.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.VacationId, CalendarEventTypes.Vacation), x => x);
+                var employeeWorkoutsValues = employeeWorkoutsTask.Result.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.ChangeId, CalendarEventTypes.Workout), x => x);
+                var employeeSickLeavesValues = employeeSickLeavesTask.Result.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.SickLeaveId, CalendarEventTypes.Sickleave), x => x);
+
+                foreach (var calendar in sharepointItemsByCalendars[departmentId].Keys)
+                {
+                    var storageItems = await sharepointItemsByCalendars[departmentId][calendar];
```

Comment:
I didn't mean anything like you did. I meant that you have a nested dictionary in another dictionary - first one by departmentId and the inner one by calendar. But it is enough to skip department at all, because all calendar names are unique and to have sharepointItemsByCalendar[calendar].

Reply:
Agree. Algorithm changed.